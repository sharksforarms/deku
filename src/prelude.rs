pub use crate::{error::DekuError, BitsReader, BitsWriter, DekuRead, DekuWrite};
pub use bitvec::{
    order::BitOrder, order::Lsb0, order::Msb0, slice::AsBits, slice::BitSlice, vec::BitVec,
};
