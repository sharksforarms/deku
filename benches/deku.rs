use criterion::{black_box, criterion_group, criterion_main, Criterion};
use deku::prelude::*;

#[derive(Debug, PartialEq, DekuRead)]
struct DekuByte {
    data: u8
}

#[derive(Debug, PartialEq, DekuRead)]
#[deku(id_type = "u8")]
enum DekuEnum {
    #[deku(id = "0x01")]
    VariantA(u8),
}

#[derive(Debug, PartialEq, DekuRead)]
struct DekuVec {
    count: u8,
    #[deku(len = "count")]
    data: Vec<u8>
}

fn deku_read_byte(input: &[u8]) {
    let (_rest, v) = DekuByte::from_bytes((input, 0)).unwrap();
    assert_eq!(DekuByte { data: 0x01 }, v);
}

fn deku_read_enum(input: &[u8]) {
    let (_rest, v) = DekuEnum::from_bytes((input, 0)).unwrap();
    assert_eq!(DekuEnum::VariantA(0x02), v);
}

fn deku_read_vec(input: &[u8]) {
    let (_rest, v) = DekuVec::from_bytes((input, 0)).unwrap();
    assert_eq!(100, v.count);
    assert_eq!(100, v.data.len());
}

fn criterion_benchmark(c: &mut Criterion) {
    c.bench_function("deku_read_byte", |b| {
        b.iter(|| deku_read_byte(black_box([0x01].as_ref())))
    });

    c.bench_function("deku_read_enum", |b| {
        b.iter(|| deku_read_enum(black_box([0x01, 0x02].as_ref())))
    });

    let deku_read_vec_input = {
        let mut v = [0xFFu8; 101].to_vec();
        v[0] = 100u8;
        v
    };
    c.bench_function("deku_read_vec", |b| {
        b.iter(|| deku_read_vec(black_box(&deku_read_vec_input)))
    });
}

criterion_group!(benches, criterion_benchmark);
criterion_main!(benches);
