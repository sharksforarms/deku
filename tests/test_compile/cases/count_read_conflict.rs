use deku::prelude::*;

#[derive(DekuRead)]
struct Test1 {
    #[deku(bits_read = "1", bytes_read = "2")]
    a: Vec<u8>,
}

#[derive(DekuRead)]
struct Test2 {
    #[deku(count = "1", bits_read = "1", bytes_read = "2")]
    a: Vec<u8>,
}

#[derive(DekuRead)]
struct Test3 {
    #[deku(count = "1", bits_read = "1")]
    a: Vec<u8>,
}

#[derive(DekuRead)]
struct Test4 {
    #[deku(count = "1", bytes_read = "2")]
    a: Vec<u8>,
}

#[derive(DekuRead)]
struct Test5 {
    #[deku(count = "1", read_all)]
    a: Vec<u8>,
}

#[derive(DekuRead)]
struct Test6 {
    #[deku(bytes_read = "1", read_all)]
    a: Vec<u8>,
}

#[derive(DekuRead)]
struct Test7 {
    #[deku(bits_read = "1", read_all)]
    a: Vec<u8>,
}

fn main() {}
