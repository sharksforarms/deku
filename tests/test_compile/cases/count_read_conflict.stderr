error: conflicting: both `bits_read` and `bytes_read` specified on field
 --> $DIR/count_read_conflict.rs:5:24
  |
5 |     #[deku(bits_read = "1", bytes_read = "2")]
  |                        ^^^

error: conflicting: both `bits_read` and `bytes_read` specified on field
  --> $DIR/count_read_conflict.rs:11:37
   |
11 |     #[deku(count = "1", bits_read = "1", bytes_read = "2")]
   |                                     ^^^

error: conflicting: both `count` and `bits_read` specified on field
  --> $DIR/count_read_conflict.rs:17:20
   |
17 |     #[deku(count = "1", bits_read = "1")]
   |                    ^^^

error: conflicting: both `count` and `bytes_read` specified on field
  --> $DIR/count_read_conflict.rs:23:20
   |
23 |     #[deku(count = "1", bytes_read = "2")]
   |                    ^^^

error: conflicting: `read_all` cannot be used with `until`, count`, `bits_read`, or `bytes_read`
  --> $DIR/count_read_conflict.rs:27:10
   |
27 | #[derive(DekuRead)]
   |          ^^^^^^^^
   |
   = note: this error originates in the derive macro `DekuRead` (in Nightly builds, run with -Z macro-backtrace for more info)

error: conflicting: `read_all` cannot be used with `until`, count`, `bits_read`, or `bytes_read`
  --> $DIR/count_read_conflict.rs:33:10
   |
33 | #[derive(DekuRead)]
   |          ^^^^^^^^
   |
   = note: this error originates in the derive macro `DekuRead` (in Nightly builds, run with -Z macro-backtrace for more info)

error: conflicting: `read_all` cannot be used with `until`, count`, `bits_read`, or `bytes_read`
  --> $DIR/count_read_conflict.rs:39:10
   |
39 | #[derive(DekuRead)]
   |          ^^^^^^^^
   |
   = note: this error originates in the derive macro `DekuRead` (in Nightly builds, run with -Z macro-backtrace for more info)
